diff --git a/srcs/CMakeLists.txt b/srcs/CMakeLists.txt
index b3bed09..eb60e43 100644
--- a/srcs/CMakeLists.txt
+++ b/srcs/CMakeLists.txt
@@ -33,13 +33,13 @@ endif()
 
 
 #verify which type of allocator we can use.
-try_compile(COMPILE_SUCCEEDED ${CMAKE_BINARY_DIR}/compile_tests SOURCES ${PROJECT_SOURCE_DIR}/../build/allocator_test.cpp CMAKE_FLAGS CXX_STANDARD 11)
-if(COMPILE_SUCCEEDED)
-  message("Allocator test passed")
-else()
-  message("Allocator test failed, enable fix")
+#try_compile(COMPILE_SUCCEEDED ${CMAKE_BINARY_DIR}/compile_tests SOURCES ${PROJECT_SOURCE_DIR}/../build/allocator_test.cpp CMAKE_FLAGS CXX_STANDARD 11)
+#if(COMPILE_SUCCEEDED)
+#  message("Allocator test passed")
+#else()
+#  message("Allocator test failed, enable fix")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DHEIF_GCC_ALLOCATOR_FIX=1")
-endif()
+#endif()
 
 
 # picked from http://stackoverflow.com/a/3818084
