diff --git a/CMakeLists.txt b/CMakeLists.txt
index a7371cb..5293a5d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,7 +35,7 @@ ecm_setup_version(PROJECT VARIABLE_PREFIX KCALCORE
 ########### Find packages ###########
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Core Gui)
 
-set(LibIcal_MIN_VERSION "2.0")
+set(LibIcal_MIN_VERSION "0.43")
 find_package(LibIcal ${LibIcal_MIN_VERSION})
 set_package_properties(LibIcal PROPERTIES TYPE REQUIRED)
 if(LibIcal_VERSION VERSION_GREATER "2.98")
diff --git a/src/icalformat_p.cpp b/src/icalformat_p.cpp
index 8708e26..b994e06 100644
--- a/src/icalformat_p.cpp
+++ b/src/icalformat_p.cpp
@@ -886,7 +886,7 @@ icalproperty *ICalFormatImpl::writeAttachment(const Attachment::Ptr &att)
     if (att->isUri()) {
         attach = icalattach_new_from_url(att->uri().toUtf8().data());
     } else {
-        attach = icalattach_new_from_data((const char *)att->data().constData(), nullptr, nullptr);
+        attach = icalattach_new_from_data((unsigned char *)att->data().constData(), nullptr, nullptr);
     }
     icalproperty *p = icalproperty_new_attach(attach);
 
