diff --git a/Source/cmake/OptionsQt.cmake b/Source/cmake/OptionsQt.cmake
index 003fdca41..f0db91f77 100644
--- a/Source/cmake/OptionsQt.cmake
+++ b/Source/cmake/OptionsQt.cmake
@@ -501,7 +501,7 @@ if (MACOS_FORCE_SYSTEM_XML_LIBRARIES)
         set(LIBXSLT_LIBRARIES xslt)
     endif ()
 else ()
-    find_package(LibXml2 2.8.0 REQUIRED)
+    find_package(LibXml2 2.7.0 REQUIRED)
     if (ENABLE_XSLT)
         find_package(LibXslt 1.1.7 REQUIRED)
     endif ()
