diff --git a/src/platformsupport/services/genericunix/qgenericunixservices.cpp b/src/platformsupport/services/genericunix/qgenericunixservices.cpp
index cdb5d33859..5a2bd5f724 100644
--- a/src/platformsupport/services/genericunix/qgenericunixservices.cpp
+++ b/src/platformsupport/services/genericunix/qgenericunixservices.cpp
@@ -146,7 +146,29 @@ static inline bool launch(const QString &launcher, const QUrl &url)
 #if !QT_CONFIG(process)
     const bool ok = ::system(qPrintable(command + QLatin1String(" &")));
 #else
-    const bool ok = QProcess::startDetached(command);
+    QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
+
+    // If we are running into AppImage bundle, switch env var to the right values.
+    if (env.contains(QLatin1String("APPIMAGE_ORIGINAL_LD_LIBRARY_PATH")) &&
+        env.contains(QLatin1String("APPIMAGE_ORIGINAL_QT_PLUGIN_PATH"))  &&
+        env.contains(QLatin1String("APPIMAGE_ORIGINAL_XDG_DATA_DIRS"))   &&
+        env.contains(QLatin1String("APPIMAGE_ORIGINAL_PATH")))
+    {
+        qCDebug(DIGIKAM_GENERAL_LOG) << "Adjusting environment variables for AppImage bundle";
+
+        env.insert(QLatin1String("LD_LIBRARY_PATH"),
+                   env.value(QLatin1String("APPIMAGE_ORIGINAL_LD_LIBRARY_PATH")));
+        env.insert(QLatin1String("QT_PLUGIN_PATH"),
+                   env.value(QLatin1String("APPIMAGE_ORIGINAL_QT_PLUGIN_PATH")));
+        env.insert(QLatin1String("XDG_DATA_DIRS"),
+                   env.value(QLatin1String("APPIMAGE_ORIGINAL_XDG_DATA_DIRS")));
+        env.insert(QLatin1String("PATH"),
+                   env.value(QLatin1String("APPIMAGE_ORIGINAL_PATH")));
+    }
+
+    QProcess process;
+    process.setProcessEnvironment(env);
+    const bool ok = process.startDetached(command);
 #endif
     if (!ok)
         qWarning("Launch failed (%s)", qPrintable(command));
