diff --git a/src/solid/devices/backends/iokit/cfhelper.cpp b/src/solid/devices/backends/iokit/cfhelper.cpp
index 9272c30..8c94737 100644
--- a/src/solid/devices/backends/iokit/cfhelper.cpp
+++ b/src/solid/devices/backends/iokit/cfhelper.cpp
@@ -94,7 +94,7 @@ static QVariant q_toVariant(const CFTypeRef &obj)
 
     if (typeId == CFDateGetTypeID()) {
         QDateTime dt;
-        dt.setTime_t(uint(kCFAbsoluteTimeIntervalSince1970));
+        dt.setMSecsSinceEpoch(uint(kCFAbsoluteTimeIntervalSince1970));
         return dt.addSecs(int(CFDateGetAbsoluteTime(static_cast<const CFDateRef>(obj))));
     }
 
