diff --git a/CMakeLists.txt b/CMakeLists.txt
index b2bdb35..733bc95 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -157,8 +157,8 @@ set (MGWHELP_IMPLIB ${CMAKE_ARCHIVE_OUTPUT_DIRECTORY}/libmgwhelp.a)
 set (EXCHNDL_IMPLIB ${CMAKE_ARCHIVE_OUTPUT_DIRECTORY}/libexchndl.a)
 
 add_subdirectory (src)
-add_subdirectory (sample)
-add_subdirectory (tests)
+#add_subdirectory (sample)
+#add_subdirectory (tests)
 
 
 ##############################################################################
diff --git a/thirdparty/zlib/CMakeLists.txt b/thirdparty/zlib/CMakeLists.txt
index bf66809..c2f3ccf 100644
--- a/thirdparty/zlib/CMakeLists.txt
+++ b/thirdparty/zlib/CMakeLists.txt
@@ -5,7 +5,7 @@ else ()
     add_definitions (-DHAVE_UNISTD_H)
 endif ()
 
-add_library (z EXCLUDE_FROM_ALL
+add_library (z STATIC
     adler32.c
     compress.c
     crc32.c
