diff --git a/.qmake.conf b/.qmake.conf
index 08c6279..4878ce4 100644
--- a/.qmake.conf
+++ b/.qmake.conf
@@ -12,3 +12,8 @@ macx:isEqual(QT_MAJOR_VERSION,5):greaterThan(QT_MINOR_VERSION, 3): CONFIG *= c++
 android: CONFIG*=c++11
 QMAKE_MACOSX_DEPLOYMENT_TARGET = 10.8
 QMAKE_IOS_DEPLOYMENT_TARGET = 6.0
+
+CONFIG += no_examples no_rpath
+INCLUDEPATH += /opt/digikam/include/
+LIBS += -L/opt/digikam/lib/
+
diff --git a/src/libQtAV.pro b/src/libQtAV.pro
index 27c1570..d80771e 100644
--- a/src/libQtAV.pro
+++ b/src/libQtAV.pro
@@ -323,7 +323,7 @@ mac {
   ios {
     OBJECTIVE_SOURCES *= codec/video/SurfaceInteropCVOpenGLES.mm
   } else {
-    CONFIG += config_vda
+#    CONFIG += config_vda
     SOURCES *= codec/video/SurfaceInteropIOSurface.cpp
     #SOURCES *= codec/video/SurfaceInteropCVOpenGL.cpp
     LIBS += -framework IOSurface
