diff --git a/CMakeLists.txt b/CMakeLists.txt
index e2b17f3..50db9b2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -56,7 +56,7 @@ ENDIF()
 
 # overwrite some folders before GNUInstallDirs is loaded
 IF (WIN32)
-    SET(CMAKE_INSTALL_DATAROOTDIR "$ENV{ALLUSERSPROFILE}" CACHE PATH "read-only architecture-independent data root (share)")
+#    SET(CMAKE_INSTALL_DATAROOTDIR "$ENV{ALLUSERSPROFILE}" CACHE PATH "read-only architecture-independent data root (share)")
 ENDIF()
 
 # get standard GNU install directories for the current platform
